### DISCUSSION TITLE
Compatible Php 8.1 Pdf repositories

### DISCUSSION THREAD

**Discussion by rossaddison** [yiisoft/demo#486] [Q&A] [answered]
I get this error using mpdf/mpdf using rossaddison/yii-invoice 

```
PHP Compile Error — Yiisoft\ErrorHandler\Exception\ErrorException
Declaration of Mpdf\Mpdf::setLogger(Psr\Log\LoggerInterface $logger) must be compatible with Psr\Log\LoggerAwareInterface::setLogger(Psr\Log\LoggerInterface $logger): void
```

and I appreciate that mpdf/mpdf is trying to accomodate php version 5.4 without type declarations at the end of their setLogger functions and so the 'void' causes this conflict and I can hack their vendor files by inserting return types, specifically 'void' at the end of their functions.  I have tried creating a fork of mpdf/mpdf including these 'hacks', for want of a better word, but composer is returning the following error.

```
 Problem 1
    - Root composer.json requires rossaddison/mpdf, it could not be found in any version, there may be a typo in the package name.

Potential causes:
 - A typo in the package name
 - The package is not available in a stable-enough version according to your minimum-stability setting
   see <https://getcomposer.org/doc/04-schema.md#minimum-stability> for more details.
 - It's a private package and you forgot to add a custom repository to find it
```

Do you know of any good php pdf generators that are strict return type focused using psr/log version 3 that will avoid this type of error?

**Comment by xepozz** [comment, marked as answer]
You need to register your own fork in `packagist.org` or specify it in composer.json in `repositories` section.
https://stackoverflow.com/a/13500676

**Reply by rossaddison** [reply]
1. Used following code to build rossaddison/mpdf         composer.json

```
{
    "name": "rossaddison/mpdf",
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/rossaddison/mpdf"
        }
    ],
    "require": {
        "mpdf/mpdf": "dev-master"
    }
}
```

I renamed the mpdf/mpdf 'development' branch, which I inherited when I forked, to 'master' so I have one branch now called master that belongs to the rossaddion/mpdf repo.  

Now because my branch is called master I can change the require section, as recommended in the link provided, from:

  ```
"require": {
        "mpdf/mpdf": "development"
    }
```

to:

  ```
"require": {
        "mpdf/mpdf": "dev-master"
    } 
```
2. Registered on packagist.org

**Reply by rossaddison** [reply]
I have had to modify the rossaddison/yii-invoice composer.json, which I like to keep in line with the yiisoft/demo composer.json, changing 

the psr requirement from:

"psr/log": "^3.0",

to: 

"psr/log": "^1.0 || ^2.0 || ^3.0",

Psr version 2 was used with mpdf 8.0.16    ie.

 ```
- Upgrading mpdf/mpdf (v8.0.16 => dev-master e511e89)
 - Downgrading psr/log (3.0.0 => 2.0.0) 
```

This confirms that the latest version of mpdf/mpdf is not compatible with psr version 3 because of stricter return type requirement on the setLogger interface using 'void'.  Hopefully one day mpdf/mpdf will drop php 5.4 to php 7.0 requirement. The tests on 7.1 to 8.1 are working. Or provide a separate branch that is compatible with psr3.

**Reply by xepozz** [reply]
> Used following code to build rossaddison/mpdf composer.json

No. You need to remove the `repositories` section from it and move it into the repo when you want to use your own fork. It's `yii-invoice` as I see.
So
```php
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/rossaddison/mpdf"
        }
    ],
```
Should be placed in yii-invoice/composer.json.

Also you don't need to require mpdf/mdpf repo because your repo (`rossaddison/mpdf`) is already it is.
If you fork a repo, than means you will get all files from the one.
So if you want to change psr/log version to 3, just edit composer.json files in your fork.

P.S. usually devs don't register temporary forks on packagist.org. Just specify fork repo in composer.json and use it until the repo owner accept yours changes if you send it.

What do you need to do:
1. Remove fork repo.
2. Fork it again
3. Fix `psr/log: 1 | 2` to `psr/log: 1 | 2 | 3` in `composer.json`
4. Do needed changes in the codebase.
5. Make sure all test are passed.
6. Create a PR and send it to the original repo
7. In your `yii-invoce` project use `mpdf/mpdf` in `require` section and add `repositories` section.

**Comment by rossaddison** [comment]
I have simplified rossaddison/mpdf composer.json to 

{
    "name": "rossaddison/mpdf"
}

and inserted 

"rossaddison/mpdf": "*", into rossaddison/yii-invoice composer.json without adding anything to repositories section of yii-invoice composer.json 

and added:  "psr/log": "^1.0|^2.0|^3.0",  to the yii-invoice composer.json.

Pdf's are working fine. Thanks for your help.
